### ISSUE TITLE
Feature request: Layer specific purpose configuration

### ISSUE THREAD

**Issue by smile13241324** [syl20bnr/spacemacs#11229]
Currently `Purpose` comes with a minimal config from it's own layer. The default config can then be adjusted on a per user basis.

I find this problematic as many layers have specific kinds of buffers which require config changes on a per layer basis i.e. adding `Anaconda buffer` to a non `general` category for `python layer`. This requires users to maintain a rather big general purpose config or to use a minimal configuration instead.

Wouldn't it be better to have each layer provide a purpose config for its specific buffers via the post-init mechanic? Purpose-layer could expose a wider default list of purposes including `help` or `compile` and the layers could map the buffers respectively.

Would this be a good idea? Or am I getting too far into personal config territory here?

**Comment by sdwolfz** [comment]
This area is beyond my knowledge, I summon @syl20bnr and @bmag to answer your questions.

**Comment by bmag** [comment]
> Would this be a good idea? Or am I getting too far into personal config territory here?

This could be a good idea, depending on how far we take it. On the one hand, giving Anaconda's documentation buffer a `Help` purpose is pretty basic and reasonable, although it may not be the best example since other help buffers are configured as popup windows (by adding them to `popwin:special-display-config`, see `spacemacs-visual/init-popwin` for example). On the other hand, giving all Python buffers a `python` purpose is a bit too much, since then it wouldn't share the same purpose with other editable buffers - for example C and Python buffers have different purposes, which could be wrong for most users.

> Wouldn't it be better to have each layer provide a purpose config for its specific buffers via the post-init mechanic?

I think it is good approach for how layers add their own purposes, and quite easy to implement as well. See below for proposed implementation.

> Purpose-layer could expose a wider default list of purposes including `help` or `compile` and the layers could map the buffers respectively.

It isn't necessary. Layers can use purposes without pre-defining the purposes elsewhere. As long as different layers use the same purpose for buffers that have similar purposes, we should be fine. There should be a list of purposes used in Spacemacs, likely in `spacemacs-purpose/README.org`, but we can build it organically and gradually when the need arises. One caveat to remember is that purposes shouldn't have the same name as existing functions (e.g. `help` is no good, but `Help` is ok), otherwise their display can't be customized in `purpose-special-action-sequences` (due to implementation details).

----

Proposed implementation:
`window-purpose` supports "extension" configurations, which have a higher priority than default purposes and a lower priority than user-defined purposes. Leveraging these is straight-forward. For user's convenience, we can also add a layer variable for not enabling the layer's purpose configuration (only consulted during startup, doesn't disable during config-reload if initially was enabled).

In `<layer>/config.el`:
```elisp
(defvar <layer>-configure-purposes-p t
  "If non-nil, the layer will configure relevant purposes.
The layer's purpose configuration is stored as a \"purpose
extension\" with keyword `:<layer>-layer', and thus can be modified by:
  (purpose-add-extension-purposes :<layer>-layer ...)
  (purpose-remove-extension-purposes :<layer>-layer ...)
  (purpose-get-extension-configuration :<layer>-layer)
  (purpose-set-extension-configuration :<layer>-layer ...)
  (purpose-del-extension-configuration :<layer>-layer)")
```
In <layer>/packages.el
```elisp
(defun <layer>/post-init-window-purpose ()
  (when <layer>-configure-purposes-p
    (purpose-set-extension-configuration :<layer>-layer ...)))
```

To disable, layer can be configured in user's layer list as such:
```elisp
(<layer> :variables <layer>-configure-purposes-p nil)
```

For example, a hypothetical (and not so good) configuration for python layer:
In `python/config.el`:
```elisp
(defvar python-configure-purposes-p t
  "If non-nil, the layer will configure relevant purposes.
The layer's purpose configuration is stored as a \"purpose
extension\" with keyword `:python-layer', and thus can be modified by:
  (purpose-add-extension-purposes :python-layer ...)
  (purpose-remove-extension-purposes :python-layer ...)
  (purpose-get-extension-configuration :python-layer)
  (purpose-set-extension-configuration :python-layer ...)
  (purpose-del-extension-configuration :python-layer)")
```
In python/packages.el
```elisp
(defun python/post-init-window-purpose ()
  (when python-configure-purposes-p
    (purpose-set-extension-configuration
     :python-layer
     (purpose-conf :mode-purposes '((python . edit))
	  	   :name-purposes '(("*Anaconda Help*" . Help))
		   :regexp-purposes '(("^\\*Anaconda" . general))))))
```

A user can add or override a layer's configuration either by user-defined purposes or by modifying the extension purposes. This code belong in `user-config`:
```elisp
(purpose-add-user-purposes :names '(("*Anaconda Help*" . general))
                           :regexps '(("\\.hy$" . python)))
```

or:
```elisp
(purpose-add-extension-purposes :python-layer
  :names '(("*Anaconda Help*" . general))
  :regexps '(("\\.hy$" . python)))
```

To remove a layer-defined purpose, the extension needs to be modified:
```elisp
(purpose-remove-extension-purposes :python-layer :names '("*Anaconda Help*"))
```

**Comment by bmag** [comment]
@smile13241324 could you expand on your use-cases, and can you think of use-cases not covered by the proposed implementation? Do you have a list of purposes that you think we should configure in Spacemacs?

**Comment by Miciah** [comment]
Perhaps relevant, #10098 defines a "chat" purpose for erc, jabber, rcirc, and slack buffers and a "mail" purpose for gnus, mu4e, and notmuch buffers.  That PR has languished for lack of testing and interest.

**Comment by smile13241324** [comment]
@bmag I have found the list of defined purposes with `SPC r P`. I have not too much additions right now I think we should extend the list while we go along.

Right now we have:
* general
* edit
* terminal
* dired
* buffers
* search
* image
* package
* minibuf
* which-key
* popb
* popr
* helm

I would propose to add:
* Mail
* Chat
* Compile
* Help

About my use cases, normally I would use a four window layout consisting of a neotree buffer an edit buffer a terminal and a help buffer. In a second frame I normally have two windows one for a compile buffer and one for a dired buffer. When every layer has allocated its buffers to the standard purposes this layout would be kept independent of the used layer.

For example when I use the python layer, opening a new source file should be opened in the edit buffer and when I show help information via "Anaconda" it should be shown in the help buffer.

It would also be possible that someone would use a similar setup with a mail and chat window in a separate frame.

I think with the proposed implementation it should be possible full fill all the various purpose config use-cases.

**Comment by smile13241324** [comment]
@bmag I have made my first PR for this ticket. It contains some macros to provide a consistent way to set the config and to have the same behaviour in all layers using it.

In addition I have added a more detailed purpose config for the python layer mainly mapping the various lookup and anaconda buffers to the help purpose.

I have tested that locally and it did work pretty good, however I have a question where to put these macros I have made. For now I have put them into `spacemacs-purpose#funcs.el` however I have noticed that this is only loaded after the entire `packages.el` and after `post-init-xxx` functions so when I actually want to execute the macro to set the specific config it is not defined yet. To work around this I have now put the setup code into the mode specific hook but this is causing the macro itself to be put into the hook instead of its evaluated form which is causing us to evaluate it with every mode we load. Do you have a better idea of where to put these macros to make sure they are loaded or would it make sense to load the `post-init-xxx` stuff only after `funcs.el`?

**Comment by bmag** [comment]
@Miciah I kind of lost that PR among all the other PRs. It's fine as it is, and I can refactor it later to the form we're discussing here. I just want to have a more careful look at how the email clients interact with it, or get someone to test it.

@smile13241324 upon further consideration, I think we don't need the layer variable (instead the user can disable a layer's configuration of window-purpose through the `:packages` keyword in their personal layers list in the dotspacemacs file). Also, I think we don't need the macros - it's a very thin abstraction layer that doesn't help much, and I'd rather not add it. With these changes you won't have that load-order headache (and btw most macros are in `core`). I think instead of the macros, a better solution is to add documentation of how layers can configure purposes idiomatically. This documentation can go either in `spacemacs-purpose` layer's README.org, or in LAYERS.org as a subsection of "How do I ... Idiomatically" (a section that used to be in the document and we can resurrect (it's still there in master branch)). I'll give a more through review in the PR page.

**Comment by smile13241324** [comment]
I think its time to start work on this again...

**Comment by smile13241324** [comment]
I have finally started working on this again. Purpose layer now describes how purpose allocations are properly added to layers. Also it shows now how these can be overwritten in the dotfile.

I will now start to check the existing purpose bindings and bring them in the correct form. Current bindings seem to use user bindings instead of extension bindings which prevents any overwriting in the dotfile. Also the loading order is often fixed manually with hooks and `with-eval-after-load` constructs which delays loading unnecessarily. Lets replace it with a proper declaration of a `post-init-window-purpose` config.

**Comment by smile13241324** [comment]
@sdwolfz @JAremko @duianto @syl20bnr I would need your opinion about the purpose `popb` and `popr`, used for buffers getting opened by `popwin`.

While I was working on the purpose configs, I have noticed that many of the buffers are bind by `pupo` automatically. This includes all modes where there is some kind of `popwin` config i.e. help, compile, dap-logs, etc. These are either bound to `popb` if they would open below or `popr` if they would open on the buffer sides.

The question is now should we stay with this categorisation or shall we introduce more specific purposes for these buffers like help and compile.

**Comment by smile13241324** [comment]
My personal opinion is that purposes should always have a specific meaning, just being a popup window is no meaning for me just a way of how it is technically opened. So I would vote for using more specific purposes for these.

**Comment by duianto** [comment]
I don't know much/hardly anything about purposes or popwin.

Go with what seems to improve the current situation, until someone suggests something better.

**Comment by smile13241324** [comment]
I have added some purposes and will continue to do so. I think with this the initial idea behind this ticket is solved and therefore should be closed. I have also included instructions how to add new purposes to layers if others want to continue from here.
